File: shared/domain.ts
// Single source of truth for domain shapes. Zod schemas validate external input on the backend;
// the frontend only imports the inferred types (`import type`), so zod is not bundled client-side.
import { z } from 'zod'

const date = z.string().regex(/^\d{4}-\d{2}-\d{2}$/, 'Expected YYYY-MM-DD')
const time = z.string().regex(/^([01]\d|2[0-3]):[0-5]\d$/, 'Expected HH:mm')

export const prioritySchema = z.enum(['low', 'medium', 'high', 'critical'])
export type Priority = z.infer<typeof prioritySchema>

const calendarEventBase = z.object({
  id: z.string().min(1).max(64),
  title: z.string().min(1).max(120),
  date, // first occurrence
  start: time,
  end: time,
  weekly: z.boolean().default(false), // repeats every week on the same weekday
  remote: z.boolean().optional(), // online / from home: no travel is reserved for it
  // 'travel' is time the user places by hand where it is hard to do other things (commute, waiting…); it replaces the
  // automatic commute on that day.
  kind: z.enum(['commitment', 'travel']).optional(),
  canOverlap: z.boolean().optional(), // other things can be done during it (reading on a train): activities that allow it may overlap
    location: z.string().max(120).optional(),
})
const endAfterStart = { message: 'end must be after start', path: ['end'] }
export const calendarEventSchema = calendarEventBase.refine((e) => e.end > e.start, endAfterStart)
export const calendarEventDraftSchema = calendarEventBase.omit({ id: true }).refine((e) => e.end > e.start, endAfterStart)
export type CalendarEvent = z.infer<typeof calendarEventSchema>

export const activitySchema = z.object({
  id: z.string().min(1).max(64),
  name: z.string().min(1).max(80),
  sessionsPerWeek: z.number().int().min(1).max(14),
  sessionMinutes: z.number().int().min(15).max(480),
  minSessionMinutes: z.number().int().min(15).max(480).optional(), // hard minimum for a useful session
  priority: prioritySchema,
  preferredDays: z.array(z.number().int().min(0).max(6)).max(7).default([]),
  preferredStart: time.optional(),
  preferredEnd: time.optional(),
  // The session may be split into blocks of this many minutes, placed at different times of the same day.
  splitMinutes: z.number().int().min(15).max(240).optional(),
  canOverlap: z.boolean().optional(), // can be done at the same time as something else (e.g. during a train ride)
  overlapWith: z.array(z.string().min(1).max(60)).max(10).optional(), // only during these events/travel (by title); empty = any
  maxOverlapMinutes: z.number().int().min(15).max(2400).optional(), // at most this much per week done on overlapped time
  onlyPreferred: z.boolean().optional(), // hard: only inside the preferred hours instead of merely preferring them
  deadline: date.optional(),
  goalId: z.string().max(64).optional(),
})
export type Activity = z.infer<typeof activitySchema>

export const goalSchema = z.object({
  id: z.string().min(1).max(64),
  title: z.string().min(1).max(120),
  deadline: date,
  priority: prioritySchema,
})
export type Goal = z.infer<typeof goalSchema>

export const transportModeSchema = z.enum(['walk', 'bike', 'bus', 'train', 'car'])
export type TransportMode = z.infer<typeof transportModeSchema>

// Daily round-trip time lost in transport, by one or more modes (e.g. walk + train). The engine reserves it before the
// first / after the last fixed commitment. Data saved with the former single `mode` is upgraded on read.
export const commuteSchema = z.preprocess(
  (v) => {
    if (v && typeof v === 'object' && 'mode' in v && !('modes' in v)) {
      const { mode, ...rest } = v as { mode: unknown }
      return { ...rest, modes: [mode] }
    }
    return v
  },
  z.object({
    modes: z.array(transportModeSchema).min(1).max(5),
    minutesPerDay: z.number().int().min(0).max(300),
  }),
)
export type Commute = z.infer<typeof commuteSchema>

// Things that must happen every day regardless of the plan (meals, medication, a walk with the dog…). They sit on top of
// free time: the engine reserves them after fixed events and travel, and never plans over them.
export const essentialSchema = z
  .object({
    id: z.string().min(1).max(64),
    title: z.string().min(1).max(60),
    start: time,
    end: time,
    kind: z.enum(['meal', 'other']),
  })
  .refine((e) => e.end > e.start, { message: 'end must be after start', path: ['end'] })
export type Essential = z.infer<typeof essentialSchema>

export const preferencesSchema = z.object({
  timezone: z.string().min(1).max(64),
  dayStart: time, // start of the waking day; everything outside [dayStart, dayEnd] is sleep/unavailable
  dayEnd: time,
  minBreakMinutes: z.number().int().min(0).max(120),
  maxDailyPlannedMinutes: z.number().int().min(60).max(960),
  commute: commuteSchema.optional(),
  essentials: z.array(essentialSchema).max(12).optional(),
  setupDone: z.boolean().optional(), // the guided setup (fixed schedule, sleep, transport, activities) was finished or skipped
})
export type Preferences = z.infer<typeof preferencesSchema>

export const reasonCodeSchema = z.enum([
  'HIGH_PRIORITY',
  'PREFERRED_TIME',
  'PREFERRED_DAY',
  'BEFORE_DEADLINE',
  'SPREAD_OUT',
  'LIGHT_DAY',
  'SHORTENED',
  'SPLIT_OVER_DAY',
  'DURING_TRAVEL',
])
export type ReasonCode = z.infer<typeof reasonCodeSchema>

export const scheduledItemSchema = z.object({
  id: z.string().max(128),
  activityId: z.string().max(64),
  date,
  start: time,
  end: time,
  reasons: z.array(reasonCodeSchema).max(12),
})
export type ScheduledItem = z.infer<typeof scheduledItemSchema>

export const planningInputSchema = z.object({
  today: date,
  weekStart: date,
  events: z.array(calendarEventSchema).max(500),
  activities: z.array(activitySchema).max(100),
  goals: z.array(goalSchema).max(100),
  preferences: preferencesSchema,
  // The plan being revised: sessions that are still valid are kept, only the affected ones move (stable replanning).
  previousItems: z.array(scheduledItemSchema).max(500).optional(),
  // Minutes since midnight of `today` at request time: what has already elapsed today is not free time any more.
  nowMinutes: z.number().int().min(0).max(1439).optional(),
})
export type PlanningInput = z.infer<typeof planningInputSchema>

// ---- Planning output ----


export type ConflictCode = 'INSUFFICIENT_AVAILABLE_TIME' | 'DEADLINE_UNACHIEVABLE'

export interface Conflict {
  code: ConflictCode
  activityId: string
  requestedMinutes: number
  scheduledMinutes: number
  availableInWindowMinutes: number // free time that existed in the allowed window before planning
  deadline?: string
}

export interface Warning {
  code: 'OVERLAPPING_EVENTS'
  detail: string
}

export type Feasibility = 'fully_feasible' | 'partially_feasible' | 'infeasible'

export interface CommuteBlock {
  date: string
  start: string
  end: string
  modes: TransportMode[]
  overlappable?: boolean // reading or listening is possible (bus, train), so compatible activities may overlap it
}

export interface EssentialBlock {
  date: string
  start: string
  end: string
  title: string
  kind: Essential['kind']
}

// What a stable replan did to the previous plan, so the change can be shown and explained.
export interface PlanChanges {
  kept: number
  moved: { from: ScheduledItem; to: ScheduledItem }[]
  dropped: ScheduledItem[] // could not be placed again
  added: ScheduledItem[] // new sessions with no previous counterpart
}

export interface PlanningResult {
  engineVersion: string
  status: Feasibility
  requestedMinutes: number
  scheduledMinutes: number
  unscheduledMinutes: number
  scheduledItems: ScheduledItem[]
  conflicts: Conflict[]
  warnings: Warning[]
  availableMinutesByDay: Record<string, number> // free before planning: waking hours minus events, commute and elapsed time
  freeMinutesByDay: Record<string, number> // still free after the planned sessions
  commuteBlocks?: CommuteBlock[]
  essentialBlocks?: EssentialBlock[]
  changes?: PlanChanges // only when the plan was built from previousItems
}

export type PlanStatus = 'proposed' | 'accepted' | 'modified'

export interface Plan {
  id: string
  weekStart: string
  status: PlanStatus
  version: number
  createdAt: string
  result: PlanningResult
}

// ---- Assistant ----

export type ProposedAction =
  // Upserts: if the id already exists in the user's data, the client updates it instead of duplicating it.
  | { type: 'create_activity'; summary: string; payload: Activity }
  | { type: 'create_goal'; summary: string; payload: Goal }
  | { type: 'create_event'; summary: string; payload: CalendarEvent }

export interface AssistantReply {
  reply: string
  proposals: ProposedAction[]
  plan?: PlanningResult // set when the assistant ran the planning engine
  planAdoptable?: boolean // false when the plan includes hypothetical (unapplied) activities
  planWeekStart?: string // the week `plan` is for (this week or the next one)
}

export const assistantRequestSchema = z.object({
  message: z.string().min(1).max(1000),
  state: planningInputSchema,
  currentItems: z.array(scheduledItemSchema).max(500).optional(), // the plan the user is looking at

})
export type AssistantRequest = z.infer<typeof assistantRequestSchema>

// ---- Accounts and stored user data ----

export const credentialsSchema = z.object({
  email: z.string().trim().toLowerCase().email().max(254),
  password: z.string().min(8, 'A palavra-passe deve ter pelo menos 8 caracteres').max(128),
})
export type Credentials = z.infer<typeof credentialsSchema>

export interface User {
  id: string
  email: string
}

// What the server stores per user. Plans and chat are opaque here (they are outputs the client already
// received from the engine/assistant); the editable inputs are validated strictly.
export const userDataSchema = z.object({
  events: z.array(calendarEventSchema).max(500),
  activities: z.array(activitySchema).max(100),
  goals: z.array(goalSchema).max(100),
  preferences: preferencesSchema,
  plans: z.record(z.string(), z.unknown()),
  chat: z.array(z.unknown()).max(60),
})
export type UserData = z.infer<typeof userDataSchema>

File: ios/Timex/Models/Domain.swift
import Foundation

// Mirrors shared/domain.ts. Every field of the server schema is listed, even the ones this app does not edit:
// PUT /api/state replaces the whole document, so a field missing here would be erased for the web client too.

enum Priority: String, Codable, CaseIterable, Identifiable {
    case low, medium, high, critical
    var id: String { rawValue }
    var label: String {
        switch self {
        case .low: "Baixa"
        case .medium: "Média"
        case .high: "Alta"
        case .critical: "Crítica"
        }
    }
}

struct CalendarEvent: Codable, Identifiable, Hashable {
    var id: String
    var title: String
    var date: String // first occurrence, YYYY-MM-DD
    var start: String // HH:mm
    var end: String
    var weekly: Bool
    var location: String?
    var remote: Bool?
    var kind: String? // "commitment" | "travel"
    var canOverlap: Bool?

    var isTravel: Bool { kind == "travel" }
}

extension CalendarEvent {
    enum CodingKeys: String, CodingKey { case id, title, date, start, end, weekly, location, remote, kind, canOverlap }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        title = try c.decode(String.self, forKey: .title)
        date = try c.decode(String.self, forKey: .date)
        start = try c.decode(String.self, forKey: .start)
        end = try c.decode(String.self, forKey: .end)
        weekly = try c.decodeIfPresent(Bool.self, forKey: .weekly) ?? false
        location = try c.decodeIfPresent(String.self, forKey: .location)
        remote = try c.decodeIfPresent(Bool.self, forKey: .remote)
        kind = try c.decodeIfPresent(String.self, forKey: .kind)
        canOverlap = try c.decodeIfPresent(Bool.self, forKey: .canOverlap)
    }
}

struct Activity: Codable, Identifiable, Hashable {
    var id: String
    var name: String
    var sessionsPerWeek: Int
    var sessionMinutes: Int
    var minSessionMinutes: Int?
    var priority: Priority
    var preferredDays: [Int] // 0 = Monday
    var preferredStart: String?
    var preferredEnd: String?
    var splitMinutes: Int?
    var canOverlap: Bool?
    var overlapWith: [String]?
    var maxOverlapMinutes: Int?
    var onlyPreferred: Bool?
    var deadline: String?
    var goalId: String?
}

extension Activity {
    enum CodingKeys: String, CodingKey {
        case id, name, sessionsPerWeek, sessionMinutes, minSessionMinutes, priority, preferredDays, preferredStart, preferredEnd
        case splitMinutes, canOverlap, overlapWith, maxOverlapMinutes, onlyPreferred, deadline, goalId
    }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        sessionsPerWeek = try c.decode(Int.self, forKey: .sessionsPerWeek)
        sessionMinutes = try c.decode(Int.self, forKey: .sessionMinutes)
        minSessionMinutes = try c.decodeIfPresent(Int.self, forKey: .minSessionMinutes)
        priority = try c.decode(Priority.self, forKey: .priority)
        preferredDays = try c.decodeIfPresent([Int].self, forKey: .preferredDays) ?? []
        preferredStart = try c.decodeIfPresent(String.self, forKey: .preferredStart)
        preferredEnd = try c.decodeIfPresent(String.self, forKey: .preferredEnd)
        splitMinutes = try c.decodeIfPresent(Int.self, forKey: .splitMinutes)
        canOverlap = try c.decodeIfPresent(Bool.self, forKey: .canOverlap)
        overlapWith = try c.decodeIfPresent([String].self, forKey: .overlapWith)
        maxOverlapMinutes = try c.decodeIfPresent(Int.self, forKey: .maxOverlapMinutes)
        onlyPreferred = try c.decodeIfPresent(Bool.self, forKey: .onlyPreferred)
        deadline = try c.decodeIfPresent(String.self, forKey: .deadline)
        goalId = try c.decodeIfPresent(String.self, forKey: .goalId)
    }
}

struct Goal: Codable, Identifiable, Hashable {
    var id: String
    var title: String
    var deadline: String
    var priority: Priority
}

struct Commute: Codable, Hashable {
    var modes: [String] // walk | bike | bus | train | car
    var minutesPerDay: Int

    enum CodingKeys: String, CodingKey { case modes, mode, minutesPerDay }
    init(modes: [String], minutesPerDay: Int) {
        self.modes = modes
        self.minutesPerDay = minutesPerDay
    }
    // Data saved by older versions has a single `mode`.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if let modes = try c.decodeIfPresent([String].self, forKey: .modes), !modes.isEmpty {
            self.modes = modes
        } else {
            self.modes = [try c.decodeIfPresent(String.self, forKey: .mode) ?? "bus"]
        }
        minutesPerDay = try c.decode(Int.self, forKey: .minutesPerDay)
    }
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(modes, forKey: .modes)
        try c.encode(minutesPerDay, forKey: .minutesPerDay)
    }
}

struct Essential: Codable, Identifiable, Hashable {
    var id: String
    var title: String
    var start: String
    var end: String
    var kind: String // meal | other
}

struct Preferences: Codable, Hashable {
    var timezone: String
    var dayStart: String
    var dayEnd: String
    var minBreakMinutes: Int
    var maxDailyPlannedMinutes: Int
    var commute: Commute?
    var essentials: [Essential]?
    var setupDone: Bool?

    static func makeDefault() -> Preferences {
        Preferences(
            timezone: TimeZone.current.identifier, dayStart: "08:00", dayEnd: "22:00", minBreakMinutes: 15, maxDailyPlannedMinutes: 480,
            commute: nil,
            essentials: [
                Essential(id: "ess-almoco", title: "Almoço", start: "12:30", end: "13:30", kind: "meal"),
                Essential(id: "ess-jantar", title: "Jantar", start: "19:30", end: "20:30", kind: "meal"),
            ],
            setupDone: nil)
    }
}

struct ScheduledItem: Codable, Identifiable, Hashable {
    var id: String
    var activityId: String
    var date: String
    var start: String
    var end: String
    var reasons: [String]
}

// ---- Planning output ----

struct Conflict: Codable, Hashable {
    var code: String
    var activityId: String
    var requestedMinutes: Int
    var scheduledMinutes: Int
    var availableInWindowMinutes: Int
    var deadline: String?
}

struct PlanWarning: Codable, Hashable {
    var code: String
    var detail: String
}

struct CommuteBlock: Codable, Hashable {
    var date: String
    var start: String
    var end: String
    var modes: [String]
    var overlappable: Bool?

    enum CodingKeys: String, CodingKey { case date, start, end, modes, mode, overlappable }
    // Plans stored by older versions have a single `mode`.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        date = try c.decode(String.self, forKey: .date)
        start = try c.decode(String.self, forKey: .start)
        end = try c.decode(String.self, forKey: .end)
        if let modes = try c.decodeIfPresent([String].self, forKey: .modes), !modes.isEmpty {
            self.modes = modes
        } else {
            self.modes = [try c.decodeIfPresent(String.self, forKey: .mode) ?? "bus"]
        }
        overlappable = try c.decodeIfPresent(Bool.self, forKey: .overlappable)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(date, forKey: .date)
        try c.encode(start, forKey: .start)
        try c.encode(end, forKey: .end)
        try c.encode(modes, forKey: .modes)
        try c.encodeIfPresent(overlappable, forKey: .overlappable)
    }
}

struct EssentialBlock: Codable, Hashable {
    var date: String
    var start: String
    var end: String
    var title: String
    var kind: String
}

struct PlanChanges: Codable, Hashable {
    struct Move: Codable, Hashable { var from: ScheduledItem; var to: ScheduledItem }
    var kept: Int
    var moved: [Move]
    var dropped: [ScheduledItem]
    var added: [ScheduledItem]
}

struct PlanningResult: Codable, Hashable {
    var engineVersion: String
    var status: String // fully_feasible | partially_feasible | infeasible
    var requestedMinutes: Int
    var scheduledMinutes: Int
    var unscheduledMinutes: Int
    var scheduledItems: [ScheduledItem]
    var conflicts: [Conflict]
    var warnings: [PlanWarning]
    var availableMinutesByDay: [String: Int]
    var freeMinutesByDay: [String: Int]?
    var commuteBlocks: [CommuteBlock]?
    var essentialBlocks: [EssentialBlock]?
    var changes: PlanChanges?
}

struct Plan: Codable, Identifiable, Hashable {
    var id: String
    var weekStart: String
    var status: String // proposed | accepted | modified
    var version: Int
    var createdAt: String
    var result: PlanningResult
}

// ---- Request bodies ----

struct PlanningInput: Codable {
    var today: String
    var weekStart: String
    var events: [CalendarEvent]
    var activities: [Activity]
    var goals: [Goal]
    var preferences: Preferences
    var previousItems: [ScheduledItem]?
    var nowMinutes: Int?
}

struct User: Codable, Hashable {
    var id: String
    var email: String
}

/// What the server stores per user. `plans` and `chat` stay opaque (see JSONValue).
struct UserData: Codable {
    var events: [CalendarEvent]
    var activities: [Activity]
    var goals: [Goal]
    var preferences: Preferences
    var plans: [String: JSONValue]
    var chat: [JSONValue]

    static func empty() -> UserData {
        UserData(events: [], activities: [], goals: [], preferences: .makeDefault(), plans: [:], chat: [])
    }
}

// ---- Assistant ----

enum ProposedAction: Codable, Hashable {
    case createActivity(summary: String, payload: Activity)
    case createGoal(summary: String, payload: Goal)
    case createEvent(summary: String, payload: CalendarEvent)

    enum CodingKeys: String, CodingKey { case type, summary, payload }

    var summary: String {
        switch self {
        case .createActivity(let s, _), .createGoal(let s, _), .createEvent(let s, _): s
        }
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let summary = try c.decode(String.self, forKey: .summary)
        switch try c.decode(String.self, forKey: .type) {
        case "create_activity": self = .createActivity(summary: summary, payload: try c.decode(Activity.self, forKey: .payload))
        case "create_goal": self = .createGoal(summary: summary, payload: try c.decode(Goal.self, forKey: .payload))
        case "create_event": self = .createEvent(summary: summary, payload: try c.decode(CalendarEvent.self, forKey: .payload))
        case let other:
            throw DecodingError.dataCorruptedError(forKey: .type, in: c, debugDescription: "Unknown proposal type \(other)")
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(summary, forKey: .summary)
        switch self {
        case .createActivity(_, let p): try c.encode("create_activity", forKey: .type); try c.encode(p, forKey: .payload)
        case .createGoal(_, let p): try c.encode("create_goal", forKey: .type); try c.encode(p, forKey: .payload)
        case .createEvent(_, let p): try c.encode("create_event", forKey: .type); try c.encode(p, forKey: .payload)
        }
    }
}

struct AssistantReply: Codable {
    var reply: String
    var proposals: [ProposedAction]
    var plan: PlanningResult?
    var planAdoptable: Bool?
    var planWeekStart: String?
}

struct AssistantRequest: Codable {
    var message: String
    var state: PlanningInput
    var currentItems: [ScheduledItem]?
}

/// Same shape the web app stores in `chat`, so a conversation started on one client shows on the other.
struct ChatMessage: Codable, Identifiable, Hashable {
    var id: String
    var role: String // user | assistant
    var text: String
    var proposals: [ProposedAction]?
    var plan: PlanningResult?
    var planAdoptable: Bool?
    var handled: Bool?
    var planWeekStart: String?
    var planDecision: String? // approved | rejected
}
